Codebase Refactor Assignment

renukote opened the pull request.

renukote committed:
refactor: update image names and image imports

renukote committed:
refactor: rename components into pascal case

renukote committed:
refactor: rename components and refactor HTML

renukote committed:
refactor: refactor components to follow atomic architecture

renukote committed:
chore: sort lines ascending

renukote committed:
feat: auto sort tailwind classes using prettier

renukote committed:
feat: sort import order using prettier

renukote committed:
feat: configure eslint on src folder

renukote committed:
test: setup unit tests using React Testing Library

renukote committed:
refactor: move global.css into the styles folder

renukote committed:
refactor: move favicon into public folder

renukote committed:
chore: update gitignore

renukote commented:
Hi Ben,

I haven't fully added all the tests yet, added unit tests for two atom components, and will try to add the remaining as soon as I get time. Everything else will be the same.

BenSheridanEdwards commented on .prettierrc.js:
```diff
@@ -0,0 +1,27 @@
+  singleQuote: false,
+  tabWidth: 2,
+  trailingComma: "all",
+  plugins: [
+    "@trivago/prettier-plugin-sort-imports",
+    "prettier-plugin-tailwindcss",
+  ],
+  importOrder: [
+    "(^react$|^react/(.*)$)",
+    "(^next$|^next/(.*)$)",
+    "<THIRD_PARTY_MODULES>",
+    "@/components/*",
+    "@/pages/*",
+    "@/public/*",
+    "@/styles/*",
+    "@/data/*",
+    "^[../]",
+    "^[./]",
+    "^.+\\.css$",
+  ],
+  importOrderSortSpecifiers: true,
```
I love this - I use this in my personal projects also

BenSheridanEdwards commented on jest.config.js:
```diff
@@ -0,0 +1,19 @@
+  rootDir: "./",
+  setupFilesAfterEnv: ["<rootDir>/jest.setup.js"],
+  moduleNameMapper: {
+    "^@/components/(.*)$": "<rootDir>/src/components/$1",
+    "^@/data/(.*)$": "<rootDir>/data/$1",
```
Nice to see aliases

BenSheridanEdwards commented on src/app/components/atoms/Button/Button.test.tsx:
```diff
@@ -0,0 +1,38 @@
+  });
+
+  it("calls onClick handler when clicked", async () => {
+    const handleClick = jest.fn();
```
Minor - Would rename to include the word mock `mockOnClick` or `onClickMock`

BenSheridanEdwards commented on src/app/components/atoms/Heading/Heading.tsx:
```diff
@@ -0,0 +1,26 @@
+
+export const Heading = ({
+  additionalClassNames,
+  tagType: Tag = "h1",
```
Where is this `Tag` coming from?

BenSheridanEdwards commented on src/app/components/atoms/Heading/Heading.tsx:
```diff
@@ -0,0 +1,26 @@
+  viewType = "large",
+}: HeadingProps) => {
+  const headingFontSize =
+    viewType === "small" ? "text-[24px]" : "text-[50px] sm:text-[80px]";
```
Right so I have already skimmed and seen you have used a specific value for the font-size. 

We want to avoid this. It doesn't allow us to take advantage of the amazing rem values that Tailwind provide - `https://tailwindcss.com/docs/font-size`.

This is a main theme across the entire codebase we should talk about so I won't keep mentioning it.

But anytime you are using `-[ ]` in Tailwind, it's a mistake 90% of the time. Let's discuss.

BenSheridanEdwards commented on src/app/components/atoms/Heading/Heading.tsx:
```diff
@@ -0,0 +1,26 @@
+  text,
+  viewType = "large",
+}: HeadingProps) => {
+  const headingFontSize =
```
Now here I would say this isn't the best naming (it's a classname so should have that in the name) and you would be better served using a library `clsx` - https://www.npmjs.com/package/clsx

Let's add to discussion

BenSheridanEdwards commented on src/app/components/atoms/Image/Image.tsx:
```diff
@@ -0,0 +1,13 @@
+import NextImage, {
+  ImageProps as NextImageProps,
+  StaticImageData,
+} from "next/image";
+
+interface ImageProps extends NextImageProps {
+  alt: string;
+  src: string | StaticImageData;
+}
+
+export const Image = ({ src, alt = "image", ...rest }: ImageProps) => {
+  return <NextImage src={src} alt={alt} {...rest} />;
+};
```
Let's talk about this, not sure why alt and src wouldn't be included in NextImage.

renukote replied:
they are included but I need to define them in the props so that I can pass them to the component from a parent component. My idea to create this atom component is to be able to follow a consistent pattern for atom components and at the same time to be able to easily update the image globally at once if it's necessary

BenSheridanEdwards commented on src/app/components/molecules/ContentCard/ContentCard.tsx:
```diff
@@ -0,0 +1,32 @@
+import { Image } from "../../atoms/Image/Image";
+
+export interface ContentCardProps {
+  headImage?: string | StaticImageData;
```
I think this should be an ImageProps object. You should definitely be able to pass an alt. Also Next/Image requires the width property.

BenSheridanEdwards commented on src/app/components/organisms/ContentShowcase/ContentShowcase.tsx:
```diff
@@ -0,0 +1,38 @@
+} from "../../molecules/ContentCard/ContentCard";
+
+interface ContentShowcaseProps {
+  bottomCardDataSet: ContentCardProps[];
+  headingText: string;
+  topCardDataSet: ContentCardProps[];
```
Yes on `ContentCardProps[]` great to see.

BenSheridanEdwards commented on src/app/components/organisms/ContentShowcase/ContentShowcase.tsx:
```diff
@@ -0,0 +1,38 @@
+
+      <div className="flex flex-wrap">
+        {topCardDataSet.map((item, index) => (
+          <ContentCard key={index} {...item} />
```
Index as a key is never advised. I'd look to use something on the card object.

BenSheridanEdwards commented on src/app/components/organisms/FooterSection/FooterSection.tsx:
```diff
@@ -0,0 +1,57 @@
+import { Button } from "../../atoms/Button/Button";
+
+export const Footer = () => {
+  const [activeTab, setActiveTab] = useState<number | null>(0);
```
Let me ask you - in what scenario would this be `null`?

renukote replied:
I'm making it null when there are no active tabs when the value is 0, the first tab will be active as I'm using an index-based tab toggle. We spoke about this on the call so resolving this.

BenSheridanEdwards commented on src/app/components/organisms/FooterSection/FooterSection.tsx:
```diff
@@ -0,0 +1,57 @@
+    <footer className="lg:mt-[4rem] lg:pb-[10rem]">
+      {tabData.map(({ title, content }, index) => (
+        <div key={title} className="rounded p-2 pb-2">
+          <div
+            className="relative h-[45px] cursor-pointer"
+            onClick={() => toggleTab(index)}
```
`onClick` on a `div`?

BenSheridanEdwards replied:
Also we're using tabs in our names but there's no Tabs component in sight?

BenSheridanEdwards commented on src/app/components/organisms/FooterSection/FooterSection.tsx:
```diff
@@ -0,0 +1,57 @@
+          </div>
+
+          {activeTab === index && (
+            <p className="container mx-auto mt-5 max-w-[1237px] px-5 leading-7 text-[#a1a1a1]">
+              {content}
+            </p>
```
`p` tag for a container?

BenSheridanEdwards commented on src/app/components/organisms/Navbar/Navbar.tsx:
```diff
@@ -0,0 +1,57 @@
+
+      {isMenuOpen && (
+        <div className="flex h-[85vh] items-center sm:hidden">
+          <nav
```
Should `nav` be on the top level?

BenSheridanEdwards commented on src/app/components/organisms/Navbar/Navbar.tsx:
```diff
@@ -0,0 +1,57 @@
+        <div className="flex h-[85vh] items-center sm:hidden">
+          <nav
+            className="mt-2 flex flex-col space-y-6 px-5"
+            onClick={() => setIsMenuOpen(false)}
```
You're using your toggleMenu function is some places but not others?

BenSheridanEdwards commented on src/app/data/allPageData.ts:
```diff
@@ -1,17 +1,17 @@
-import icon10 from "../../../public/Icon_2.png";
-import icon11 from "../../../public/Frame 1321314690.png";
-import icon12 from "../../../public/bar-chart-circle-01.png";
+import icon12 from "../../../public/images/bar-chart-circle.png";
+import icon7 from "../../../public/images/cpu-chip.png";
+import icon8 from "../../../public/images/cube.png";
+import icon2 from "../../../public/images/frame-1.png";
+import icon11 from "../../../public/images/frame-2.png";
+import icon5 from "../../../public/images/frame-3.png";
+import icon1 from "../../../public/images/icon-1.png";
+import icon10 from "../../../public/images/icon-2.png";
+import icon6 from "../../../public/images/percent.png";
+import icon3 from "../../../public/images/server.png";
+import icon9 from "../../../public/images/shield-tick.png";
+import icon4 from "../../../public/images/target.png";
```
Do we really want our Icon names to be numbered?

BenSheridanEdwards commented on src/app/page.tsx:
```diff
@@ -1,19 +1,28 @@
+export default function Home() {
+  return (
+    <>
+      <ContentShowcase
```
What is the top component on a page called?

renukote replied:
since I'm using it twice at the bottom of the page, I'm not calling it as a HeroComponent, we spoke about this over the call as well. So resolving the comment

BenSheridanEdwards reviewed (commented):
I think there's more that can be done but a good first attempt. 80% of the way. Branch off, read the Tailwind docs, make improvements and open another separate PR and we can discuss both.

BenSheridanEdwards commented:
> Hi Ben,
> 
> I haven't fully added all the tests yet, added unit tests for two atom components, and will try to add the remaining as soon as I get time. Everything else will be the same.

Was happy with your tests. Happy with the structure. Focus on my comments.

BenSheridanEdwards commented on src/app/components/atoms/Heading/Heading.test.tsx:
```diff
@@ -0,0 +1,51 @@
+
+describe("Heading", () => {
+  const renderHeading = (additionalProps?: Partial<HeadingProps>) => {
+    render(<Heading {...additionalProps} text="Sample Heading" />);
```
```suggestion
    render(<Heading text="Sample Heading"  {...additionalProps} />);
```

renukote committed:
tests: implement best practices in test files

renukote committed:
This commit will be reverted

renukote committed:
Merge pull request #3 from renukote/new-branch

This commit will be reverted

renukote committed:
Revert "Demo Pull Request"

renukote committed:
Merge pull request #4 from renukote/revert-3-new-branch

Revert "Demo Pull Request"

renukote committed:
improve(feedback): refactor components to follow semantics and modularity

renukote committed:
refactor: use clsx for dynamic classes

renukote committed:
chore: syntax improvements

renukote committed:
fix: fix heading tag hierarchy

renukote committed:
fix: fix padding issues in mobile view

renukote committed:
improve: reuse child component prop types

renukote commented:
Hi @BenSheridanEdwards  

I addressed all the comments and made a few more improvements by myself. However, I am in a dilemma with the Image atom component I created. My idea is to follow a consistent atomic structure so I created an Image atom component but it's basically the same as the next/image without any additional configuration. So I'm wondering if it's really necessary at all. What do you think is the best practice in such cases?
